### ISSUE TITLE
Logout points to /gitlab/users/sign_out which is nonexistent

### ISSUE THREAD

**Issue by berndblume** [gitlabhq/gitlabhq#9387]
#6099 has been closed, but the issue persists. According to comments in the Internet, that issue has been around since a year or so. It's a major issue. How can anyone use GitLab in production if you cannot log out?

**Comment by stanhu** [comment]
It's a feature, not a bug. Once you're in GitLab, you never leave GitLab. :)

It works fine for me without `relative_url_root`, which is what most people use in production. Is `relative_url_root` configured on your system?

**Comment by berndblume** [comment]
That's my suspicion, that it's a feature! ;-)
The thing is I can't even log in as a different user if I can't log out. :-)
Regarding `relative_url_root` => I have no idea. I just installed the "omnibus" package from the gitlab.com website under Ubuntu 14.04.
Where can I look up if I have `relative_url_root` or not?

**Comment by stanhu** [comment]
Check `/var/opt/gitlab/gitlab-rails/etc/gitlab.yml`. Is `relative_url_root` commented out? What does `/var/log/gitlab/gitlab-rails/production.log` show when you attempt to logout?

**Comment by berndblume** [comment]
I'm going without `relative_url_root`, the corresponding line in `/var/opt/gitlab/gitlab-rails/etc/gitlab.yml` is commented out.

Here's what's in my `/var/log/gitlab/gitlab-rails/production.log`:
```
Started GET "/users/sign_out" for 127.0.0.1 at 2015-06-13 21:22:20 +0000
Processing by ProjectsController#show as HTML
  Parameters: {"namespace_id"=>"users", "id"=>"sign_out"}
Filter chain halted as :project rendered or redirected
Completed 404 Not Found in 31ms (Views: 22.9ms | ActiveRecord: 1.6ms)
```
Thanks in advance for any help!

**Comment by stanhu** [comment]
Thanks. I guess I was confused because the title of the issue contains `/gitlab/users/sign_out`, which suggests you are using a relative URL path to access GitLab, but it doesn't look like that's the case based on the log and your configuration file.

What is `external_url` configured in `/etc/gitlab/gitlab.rb`? If it doesn't have `http://xxx`, I wonder if that's what's causing the [`Filter chain halted`](http://stackoverflow.com/questions/22056501/overwriting-devise-after-sign-in-path-for-gives-error-filter-chain-halted-as-r) error message

**Comment by berndblume** [comment]
Oh, sorry about the confusion I created. I just copied the title of the issue from #9387, which is very similar...
Here's what my `/etc/gitlab/gitlab.rb` says: `external_url 'https://gitlab.hau.to'`
The whole GitLab installation does work, just the the logout doesn't.
To complicate things more: I did have situations in the meantime where the logout did work. But the next day I had the old thing again, logout not working!
One more thing: The "add members" button in projects and groups is not working anymore. The button is just dead, no reaction at all. These buttons used to work for a week or so, but today they don't. When I use "Inspect element" in my browser, it shows the `</button>` is not inside any `</form>`, so I'm wondering how that can be.
Feel free to register a the mentioned URL `https://gitlab.hau.to` and check it out.
Thanks again for your help!
Bernd

**Comment by stanhu** [comment]
Thanks for providing a way to see this for myself. I now can see the failure. It looks like the browser is using the wrong HTTP method for some reason. Notice the method is `GET`, not `DELETE`:

![image](https://cloud.githubusercontent.com/assets/963826/8239810/115741b0-15b3-11e5-86c9-347b7b999b12.png)

This confuses GitLab because it tries to find a project called `users`, which doesn't exist.

What's puzzling to me is that the button looks like it's set up to do the right thing:

![image](https://cloud.githubusercontent.com/assets/963826/8239798/fce2dfe6-15b2-11e5-99fb-fc2c8e979257.png)

Makes me wonder if there is some JavaScript intercepting this button and causing the wrong behavior.

**Comment by stanhu** [comment]
Actually, it looks like there is a problem rendering the main application JavaScript, which is probably why the button is doing the wrong thing:

![image](https://cloud.githubusercontent.com/assets/963826/8239842/4261aa20-15b3-11e5-80c6-cbf87a1a7b5c.png)

Are you installing from source or omnibus? You may want to recompile the assets:

https://gitlab.com/gitlab-org/gitlab-ce/blob/master/doc/raketasks/maintenance.md#precompile-the-assets

**Comment by berndblume** [comment]
Thanks so much Stan for your analysis!
I'm using the omnibus package.
I tried clearing the redis cache, it didn't help.
When I tried to recompile the assets, I got some weird EACCES errno even though I ran it as root.
```
root@water-gitlab ~ # gitlab-rake assets:precompile
rake aborted!
Errno::EACCES: Permission denied @ rb_sysopen - /opt/gitlab/embedded/service/gitlab-rails/public/assets/manifest-7b799b7a311032426750944df77f809a.json
/opt/gitlab/embedded/service/gem/ruby/2.1.0/gems/sprockets-2.11.0/lib/sprockets/manifest.rb:220:in `initialize'
/opt/gitlab/embedded/service/gem/ruby/2.1.0/gems/sprockets-2.11.0/lib/sprockets/manifest.rb:220:in `open'
/opt/gitlab/embedded/service/gem/ruby/2.1.0/gems/sprockets-2.11.0/lib/sprockets/manifest.rb:220:in `save'
/opt/gitlab/embedded/service/gem/ruby/2.1.0/gems/sprockets-2.11.0/lib/sprockets/manifest.rb:140:in `compile'
/opt/gitlab/embedded/service/gem/ruby/2.1.0/gems/sprockets-rails-2.2.4/lib/sprockets/rails/task.rb:70:in `block (3 levels) in define'
/opt/gitlab/embedded/service/gem/ruby/2.1.0/gems/sprockets-2.11.0/lib/rake/sprocketstask.rb:146:in `with_logger'
/opt/gitlab/embedded/service/gem/ruby/2.1.0/gems/sprockets-rails-2.2.4/lib/sprockets/rails/task.rb:69:in `block (2 levels) in define'
Tasks: TOP => assets:precompile
(See full trace by running task with --trace)
root@water-gitlab ~ # ls -l /opt/gitlab/embedded/service/gitlab-rails/public/assets/manifest-7b799b7a311032426750944df77f809a.json
-rw-r--r-- 1 root root 283357 May 28 14:10 /opt/gitlab/embedded/service/gitlab-rails/public/assets/manifest-7b799b7a311032426750944df77f809a.json
root@water-gitlab ~ # 
```
I'm running all this in a SmartOS LX-branded zone using the Ubuntu14.04 image provided by Joyent/SmartOS. LX-branded zones under smartOS are still in beta, so maybe it's some problem on their end. But: I still se a lot of 404 at logout problems when I google for it. Not all of these people have SmartOS LX branded zones, I guess... :-)
I will install the "real" Ubuntu14.04 under a KVM Virtual Machine on my SmartOS box, reinstall GitLab omnibus there and see if it works. I'll keep you posted.
Bernd

**Comment by stanhu** [comment]
It looks like `gitlab-rake assets:precompile` fails for me as well. We should file that as a regression.

Looking at the `strace` output of `gitlab-rake`, I believe `gitlab-rake` does a `setuid` to change the user to `git`. No wonder.

You might be able to work around the issue by doing:

```
sudo chown -R git /opt/gitlab/embedded/service/gitlab-rails/public/assets/
sudo gitlab-rake assets:precompile
```

**Comment by stanhu** [comment]
Oh, it looks like this is known behavior:

https://gitlab.com/gitlab-org/omnibus-gitlab/issues/492#note_1045405

Perhaps you can just reinstall the omnibus package and see if that helps?

**Comment by berndblume** [comment]
Oh OK. Why do they recommend the recompile on their website if it's not supposed to work anyway?
I have installed the whole omnibus package into a completely fresh machine. Unfortunately, same problem at logout...
Maybe it's really some problem with SmartOS. Their LX zones are "like" Linux, but it's not really Linux. It runs under the illumos kernel (Solaris descendant), and emulates a Linux kernel 3.13. All that is still in beta, so there might be issues there.
I'll go ahead and use KVM under SmartOS to install a real Ubuntu Linux, install the omnibus package there and see where that gets me.

**Comment by stanhu** [comment]
@berndblume It looks like it's one file in question:

```
$ md5sum /opt/gitlab/embedded/service/gitlab-rails/public/assets/application-f60a4fe80499ba2aeedc1dc58238c410.js
2167ff001f25c05283003253a7d46e90  /opt/gitlab/embedded/service/gitlab-rails/public/assets/application-f60a4fe80499ba2aeedc1dc58238c410.js
```

When I retrieve the file from your Web server, I get a different result:

```
$ curl -o bad.js "https://gitlab.hau.to/assets/application-f60a4fe80499ba2aeedc1dc58238c410.js"
$ md5sum bad.js 
89b47ec3485d27eddd4285c8511b2614  bad.js
```

**Comment by berndblume** [comment]
Oh thanks for that detective work!
That file was on my server with the correct md5.
I was using a setup with gitlab (omnibus) + gitlab nginx on one host. This was then reverse proxied into second ngnix on a different host inside my local network.
The gitlab nginx used http, and the reverse proxy nginx used SSL to the outside world.
I found out that even the gitlab nginx serves corrupted files (wrong md5) when using http. Only when it uses https it worked correctly.
So I connected the two nginx's via SSL on both of them, effectively generating two SSL encrypted connections, reverse proxied.
That made things much better, normal assets were usually served correctly this way. BUT: when I tried to download a ZIP file from a repo, these (huge) ZIP files were not served correctly and could not be unzipped. Same with tar.gz by the way.
The final solution was doing away completely with two nginx's. I'm now just using my reverse proxy (on a different host in the same subnet), then I made unicorn listen to 0.0.0.0 (instead of the default 127.0.0.1) in /etc/gitlab/gitlab.rb and connected unicorn directly as 'upstream' to my nginx (acting as reverse proxy) on the other host.
That made it finally work!
I spent hours looking for the correct nginx settings to make two nginx instances after each other work, but none of them seem to do the trick.
After all, it proved unnecessary. :-)
Thanks so much Stan for your really helpful work and support!!!
Everything is working now on my side. Without your help, I would have given up...

**Comment by stanhu** [comment]
Bernd, that's great to hear! Thanks for sharing the details of your solution. I'm a bit baffled by how the files were corrupted because if you look at the bad file, it could pass as a JavaScript bundle. 

It sounds like you quite a complex setup, but I'm glad you were able to work through all the details and make it work. I'll close out this issue unless you have any further questions. Please let us know if we can be of help in the future!

**Comment by baisoo** [comment]
change  timeout value add proxy_buffering
proxy_read_timeout      600;
    proxy_connect_timeout   600;
    proxy_buffering         off;

**Comment by maksr** [comment]
Two years after this issue was closed I had a similar problem with Gitlab and found this discussion. On sign out I got 404. In my case it happened only in Firefox. In IE and Chrome sign out worked OK. I saw in Gitlab log Firefox sends GET instead of DELETE. In javascript console I found I had some weird plugin (DOMCrypt) that somehow interfered with the js of Gitlab. When I disabled DOMCrypt plugin in Firefox, sign out in Gitlab worked fine.
